IMPRO-1506: Plugin for applying reliability calibration.

bayliffe opened the pull request:
This plugin applies reliability calibration for the simple case of a consolidated (collapsed x/y) reliability table to a forecast cube. This consolidation helps ensure a sufficient sample count in the reliability table, which avoids issues such as unpopulated bins etc. It also allows a very fast application of the calibration using ```np.interp```.

An associated notebook demonstrates how the code can be extended to work with non-consolidated reliability tables.

There are currently no new unit tests as I would appreciate a review of the simplifications assumed before writing these in case the code changes substantially.

Testing:
 - [x] Ran tests and they passed OK
 - [ ] Added new tests for the new feature(s)

benfitzpatrick commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- improver/calibration/reliability_calibration.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355&bid=16826566)

benfitzpatrick commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +490,156 @@ def process(self, cubes, coordinates=None):
         return result
+
+
+class ApplyReliabilityCalibration:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Missing class docstring](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355)

benfitzpatrick commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  3
- improver/calibration/reliability_calibration.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355&bid=16826566)

benfitzpatrick commented on improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py:
```diff
@@ -0,0 +1,307 @@
+
+from improver.calibration.reliability_calibration import (
+    ConstructReliabilityCalibrationTables as CalPlugin)
+from improver_tests.calibration.reliability_calibration.\
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Test_Setup imported from improver_tests.calibration.reliability_calibration.test_ConstructReliabilityCalibrationTables](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355)

benfitzpatrick commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  3
- improver/calibration/reliability_calibration.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355&bid=16826566)

benfitzpatrick commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/utilities/test_interpolation.py  1
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  3
- improver/utilities/interpolation.py  3
- improver/calibration/reliability_calibration.py  1
         

Clones added
============
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355&bid=16826566)

benfitzpatrick commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/utilities/test_interpolation.py  1
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  3
- improver/utilities/interpolation.py  3
- improver/calibration/reliability_calibration.py  1
         

Clones added
============
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355&bid=16826566)

fionaRust commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+        Initialise class for applying reliability calibration.
+
+        Args:
+            minimum_forecast_count (int):
```
Is this docstring correct?

bayliffe replied:
Fixed.

bayliffe commented on improver_tests/utilities/test_interpolation.py:
```diff
@@ -0,0 +1,99 @@
+from improver.utilities.interpolation import interpolate_1d
+
+
+class Test_interpolate_1d(unittest.TestCase):
```
Removed.

bayliffe commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+    def _calculate_reliability_probabilities(self, reliability_table):
+        """
+        Calculates forecast probabilities and observation frequencies from the
+        reliability table. Where the forecast count is zero, the returned
```
Fixed.

bayliffe commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+            raise ValueError('Threshold coordinates do not match between '
+                             'reliability table and forecast cube.')
+
+    def _ensure_monotonicity(self, cube):
```
I think non-monotonic probabilities against thresholds are always wrong, investigation or not. We may want to learn about how often this happens etc, but that is easily done in a notebook by patching the plugin, so I'd rather not put an option in that allows for the production of non-self-consistent output. Is that okay?

fionaRust replied:
Fair enough

bayliffe commented on improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py:
```diff
@@ -0,0 +1,343 @@
+
+    """Test the _ensure_monotonicity method."""
+
+    def test_monotonic_case(self):
```
Done.

bayliffe commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+    def _ensure_monotonicity(self, cube):
+        """
+        Ensures that probabilities change monotonically relative to thresholds
+        in the expected order, e.g. exceedance proabilities always remain the
```
Fixed.

bayliffe commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+                frequencies.
+        Returns:
+            (tuple): tuple containing:
+                **forecast_probability** (numpy.array):
```
Yup. Fixed.

bayliffe commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+
+            interpolated = interpolated.reshape(shape).astype(np.float32)
+
+            calibrated_cubes.append(forecast_threshold.copy(data=interpolated))
```
Done.

bayliffe commented on improver_tests/utilities/test_interpolation.py:
```diff
@@ -0,0 +1,99 @@
+        x = np.array([-2, 0, 3, 7, 10, 12]) / 10.
+        result = interpolate_1d(x, self.xp, self.yp, min_y=0, max_y=10)
+
+        print(result)
```
Tests removed.

bayliffe commented on improver/utilities/interpolation.py:
```diff
@@ -0,0 +1,68 @@
+    y[x < xp[0]] = (yp[0] + (x[x < xp[0]] - xp[0]) *
+                    (yp[0] - yp[1]) / (xp[0] - xp[1]))
+    y[x > xp[-1]] = (yp[-1] + (x[x > xp[-1]] - xp[-1]) *
+                     (yp[-1] - yp[-2]) / (xp[-1] - xp[-2]))
```
Moved to use this. Thanks.

bayliffe commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+
+class ApplyReliabilityCalibration:
+
+    def __init__(self, minimum_forecast_count=200):
```
Added a reference to the missing plugin docstring and referenced that here.

bayliffe commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+        # In some bins have insufficient counts, return None to avoid applying
+        # calibration.
+        valid_bins = np.where(forecast_count >= self.minimum_forecast_count)
+        if valid_bins[0].size != forecast_count.size:
```
That's not quite what this is saying. The function np.where returns a tuple of 1-d arrays, each of the same length by necessity, as all combined they give the index of each position that we want to use. As such, I cannot use a ```valid_bins, =``` construct as unless I have a 1-d input array (which is pretty much never the case here), this will cause an error (too many values to unpack). I need to take any of the arrays returned within the tuple and check the size, and zero is guaranteed to always be there.

bayliffe commented on improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py:
```diff
@@ -0,0 +1,343 @@
+
+    def test_monotonic_case(self):
+        """Test that for a probability cube in which the data is already
+        ordered monotonically it us unchanged by this method. Additionally, no
```
Fixed.

bayliffe commented on improver/calibration/reliability_calibration.py:
```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+        calibrated_cubes = iris.cube.CubeList()
+        for forecast_threshold, reliability_threshold in slices:
+            shape = forecast_threshold.shape
+            forecast_data = forecast_threshold.data.flatten()
```
Done here and in the tests.

fionaRust commented on improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py:
```diff
@@ -255,6 +259,78 @@ def test_insufficient_forecast_count(self):
+                                          self.observation_frequencies)
+
+        assert_allclose(result, expected)
+        assert_allclose(result.data, expected.data)
```
Do you need to check result.data.data and result.data.mask separately?

bayliffe replied:
Not really, but in this case I was trying to demonstrate that the interpolation is still done with whatever point is underneath the mask (valid or NaN), hence the importance of putting the mask back. I'll make it a NaN to make the point a bit more obvious (plus who knows what it will actually throw back at us; let's find out).

benfitzpatrick commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  3
- improver/calibration/reliability_calibration.py  1
         

Clones added
============
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355&bid=16826566)

benfitzpatrick commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  3
- improver/calibration/reliability_calibration.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355&bid=16826566)

fionaRust commented:
@bayliffe is Caroline's comment on the ticket (about metadata) still outstanding?

gavinevans commented on improver/calibration/reliability_calibration.py:
```diff
@@ -623,6 +634,48 @@ def _calculate_reliability_probabilities(self, reliability_table):
+                     observation_frequencies):
+        """
+        Perform interpolation of the forecast probabilities using the
+        calibration data to produce the calibrated forecast. Where necessary
```
Should `calibration data` be something like `reliability table information`?

bayliffe replied:
Done.

gavinevans commented on improver/calibration/reliability_calibration.py:
```diff
@@ -623,6 +634,48 @@ def _calculate_reliability_probabilities(self, reliability_table):
+
+        Returns:
+            numpy.ndarray:
+                The calibrated forecast probabilites. The final results are
```
Typo: `probabilites` => `probabilities`

bayliffe replied:
Done.

gavinevans commented on improver/calibration/reliability_calibration.py:
```diff
@@ -623,6 +634,48 @@ def _calculate_reliability_probabilities(self, reliability_table):
+            numpy.ndarray:
+                The calibrated forecast probabilites. The final results are
+                clipped to ensure any extrapolation has not yielded
+                probabilties outside the range 0 to 1.
```
Typo: `probabilties` => `probabilities`

bayliffe replied:
Every spelling but the right one...

benfitzpatrick commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  3
- improver/calibration/reliability_calibration.py  1
         

Clones added
============
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355&bid=16826566)

bayliffe commented:
@fionaRust  I have added Caroline's base plugin stuff. I added a print out of the calibrated cube to the notebook to demonstrate that the title is prefixed 'Post Processed' as expected.

gavinevans commented on improver/calibration/reliability_calibration.py:
```diff
@@ -494,7 +494,7 @@ def process(self, cubes, coordinates=None):
 
 
-class ApplyReliabilityCalibration:
+class ApplyReliabilityCalibration(PostProcessingPlugin):
```
Just as a query, I notice that the `AggregateReliabilityCalibrationTables` plugin isn't currently inheriting from anything. Should this plugin also be amended to inherit from `BasePlugin`?

bayliffe replied:
Added BasePlugin.

benfitzpatrick commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  3
- improver/calibration/reliability_calibration.py  1
         

Clones added
============
- improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5117355&bid=16826566)
